File: components/g8ed/services/auth/api_key_service.js
// Copyright (c) 2026 Lateralus Labs, LLC.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import crypto from 'crypto';
import { now } from '../../models/base.js';
import { ApiKeyDocument } from '../../models/auth_models.js';
import { logger } from '../../utils/logger.js';
import { ApiKeyStatus, ApiKeyError, API_KEY_LOG_PREFIX_LENGTH } from '../../constants/auth.js';
import { API_KEY_PREFIX } from '../../constants/operator_defaults.js';

/**
 * ApiKeyService - Domain Layer (Orchestration) for API Keys.
 * Handles validation, generation, and orchestration of API key lifecycle.
 * Adheres to Domain Layer principles from developer.md.
 */
class ApiKeyService {
    /**
     * @param {Object} options
     * @param {Object} options.apiKeyDataService - ApiKeyDataService instance (Data Layer)
     * @param {Object} options.internalHttpClient - InternalHttpClient instance (Relay)
     */
    constructor({ apiKeyDataService, internalHttpClient }) {
        if (!apiKeyDataService) throw new Error('ApiKeyService requires apiKeyDataService');
        this._data = apiKeyDataService;
        this._internalHttp = internalHttpClient;
        logger.info('[API-KEY-SERVICE] Domain service initialized');
    }

    /**
     * Generate a new raw API key via g8ee authority.
     * @param {string} prefix - Key prefix
     * @param {G8eHttpContext} g8eContext - REQUIRED G8eHttpContext
     * @returns {Promise<string>}
     */
    async generateRawKey(prefix = API_KEY_PREFIX, g8eContext = null) {
        if (!this._internalHttp) {
            throw new Error('InternalHttpClient is required for key generation - g8ee must be reachable');
        }

        try {
            const response = await this._internalHttp.generateApiKey(prefix, g8eContext);
            if (response.success && response.api_key) {
                return response.api_key;
            }
            throw new Error(response.error || 'Failed to generate API key via g8ee');
        } catch (error) {
            logger.error('[API-KEY-SERVICE] g8ee key generation failed', { error: error.message });
            throw new Error(`g8ee unreachable: ${error.message}. Ensure g8ee is running and accessible.`);
        }
    }

    /**
     * Validate a raw API key.
     * @param {string} apiKey
     * @param {Object} [options]
     * @param {string} [options.system_fingerprint] - Optional system fingerprint to validate against
     * @returns {Promise<{success: boolean, data?: ApiKeyDocument, error?: string}>}
     */
    async validateKey(apiKey, { system_fingerprint = null } = {}) {
        try {
            if (!apiKey) {
                return { success: false, error: 'API key is required' };
            }
            if (!apiKey.startsWith(API_KEY_PREFIX)) {
                return { success: false, error: ApiKeyError.INVALID_KEY_FORMAT };
            }

            const docId = this._data.makeDocId(apiKey);
            const doc = await this._data.getKey(docId);

            if (!doc) {
                logger.warn('[API-KEY-SERVICE] API key not found', { 
                    api_key_prefix: this._getLogPrefix(apiKey) 
                });
                return { success: false, error: 'API key not found' };
            }

            if (doc.status !== ApiKeyStatus.ACTIVE) {
                logger.warn('[API-KEY-SERVICE] API key is not active', {
                    api_key_prefix: this._getLogPrefix(apiKey),
                    status: doc.status
                });
                return { success: false, error: `API key is ${doc.status}` };
            }

            if (doc.expires_at && doc.expires_at < now()) {
                logger.warn('[API-KEY-SERVICE] API key has expired', {
                    api_key_prefix: this._getLogPrefix(apiKey),
                    expires_at: doc.expires_at
                });
                return { success: false, error: 'API key has expired' };
            }

            // Enforce fingerprint matching if established
            if (doc.system_fingerprint && system_fingerprint && doc.system_fingerprint !== system_fingerprint) {
                logger.error('[API-KEY-SERVICE] Fingerprint mismatch', {
                    api_key_prefix: this._getLogPrefix(apiKey),
                    expected: doc.system_fingerprint,
                    received: system_fingerprint
                });
                return { success: false, error: ApiKeyError.INVALID_FINGERPRINT };
            }

            return { success: true, data: doc };
        } catch (error) {
            logger.error('[API-KEY-SERVICE] Validation failed', { error: error.message });
            return { success: false, error: 'Internal validation error' };
        }
    }

    /**
     * Issue (create and store) a new API key.
     * @param {string} apiKey
     * @param {Object} keyParams
     * @returns {Promise<{success: boolean, error?: string}>}
     */
    async issueKey(apiKey, keyParams) {
        try {
            const docId = this._data.makeDocId(apiKey);
            const doc = ApiKeyDocument.parse({
                ...keyParams,
                created_at: keyParams.created_at || now(),
                status: keyParams.status || ApiKeyStatus.ACTIVE
            });

            await this._data.createKey(docId, doc);

            logger.info('[API-KEY-SERVICE] API key issued', {
                api_key_prefix: this._getLogPrefix(apiKey),
                user_id: doc.user_id
            });

            return { success: true };
        } catch (error) {
            logger.error('[API-KEY-SERVICE] Failed to issue API key', { error: error.message });
            return { success: false, error: error.message };
        }
    }

    /**
     * Update the last used timestamp of a key and establish fingerprint if missing.
     * @param {string} apiKey
     * @param {Object} [options]
     * @param {string} [options.system_fingerprint] - Optional system fingerprint to establish
     */
    async recordUsage(apiKey, { system_fingerprint = null } = {}) {
        try {
            const docId = this._data.makeDocId(apiKey);
            const doc = await this._data.getKey(docId);
            
            if (!doc) return;

            const updates = { last_used_at: now() };

            // Establish fingerprint if not already set (immutable thereafter)
            if (!doc.system_fingerprint && system_fingerprint) {
                updates.system_fingerprint = system_fingerprint;
                logger.info('[API-KEY-SERVICE] Established system fingerprint for API key', {
                    api_key_prefix: this._getLogPrefix(apiKey),
                    system_fingerprint
                });
            }

            await this._data.updateKey(docId, updates);
        } catch (error) {
            logger.warn('[API-KEY-SERVICE] Failed to record usage', { error: error.message });
        }
    }

    /**
     * Revoke (delete) an API key.
     * @param {string} apiKey
     * @returns {Promise<{success: boolean}>}
     */
    async revokeKey(apiKey) {
        try {
            const docId = this._data.makeDocId(apiKey);
            await this._data.deleteKey(docId);
            return { success: true };
        } catch (error) {
            logger.error('[API-KEY-SERVICE] Failed to revoke key', { error: error.message });
            return { success: false };
        }
    }

    _getLogPrefix(apiKey) {
        return apiKey.substring(0, API_KEY_LOG_PREFIX_LENGTH) + '...';
    }
}

export { ApiKeyService };

File: services/g8ee/app/services/auth/api_key_service.py
# Copyright (c) 2026 Lateralus Labs, LLC.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import hashlib
import logging
import secrets
from typing import TYPE_CHECKING, Any

from app.constants import DB_COLLECTION_API_KEYS, ApiKeyStatus
from app.models.api_keys import ApiKeyDocument
from app.services.cache.cache_aside import CacheAsideService
from app.utils.timestamp import now

if TYPE_CHECKING:
    from app.services.infra.settings_service import SettingsServiceProtocol

logger = logging.getLogger(__name__)

# Constants for hashing (aligned with client)
KEY_DERIVATION_ALGORITHM = "sha256"
API_KEY_HASH_LENGTH = 32

class ApiKeyService:
    """
    ApiKeyService for g8ee.
    Handles validation and lifecycle of API keys.
    """

    def __init__(self, cache_aside: CacheAsideService):
        self.cache = cache_aside
        self.collection = DB_COLLECTION_API_KEYS

    def make_doc_id(self, raw_material: str) -> str:
        """Generate a deterministic document ID from a raw API key."""
        return hashlib.sha256(raw_material.encode()).hexdigest()[:API_KEY_HASH_LENGTH]

    def generate_raw_key(self, prefix: str = "g8e_") -> str:
        """Generate a new raw API key.
        
        Default prefix is 'g8e_' matching canonical format from protocol/constants/api_key_patterns.json.
        For operator keys, use format: g8e_{8hex}_{64hex} (e.g., g8e_1a2b3c4d_...64hex...).
        For regular keys, use format: g8e_{64hex}.
        """
        return f"{prefix}{secrets.token_hex(32)}"

    async def validate_key(self, raw_key: str, system_fingerprint: str | None = None) -> tuple[bool, ApiKeyDocument | None, str | None]:
        """Validate a raw API key."""
        if not raw_key:
            return False, None, "API key is required"

        doc_id = self.make_doc_id(raw_key)
        data = await self.cache.get_document_with_cache(self.collection, doc_id)

        if not data:
            return False, None, "API key not found"

        doc = ApiKeyDocument.model_validate(data)

        if doc.status != ApiKeyStatus.ACTIVE:
            return False, doc, f"API key is {doc.status}"

        if doc.expires_at and doc.expires_at < now():
            return False, doc, "API key has expired"

        # Enforce fingerprint matching if established
        if doc.system_fingerprint and system_fingerprint and doc.system_fingerprint != system_fingerprint:
            logger.error(
                "[API-KEY-SERVICE] Fingerprint mismatch",
                extra={
                    "doc_id": doc_id[:8] + "...",
                    "expected": doc.system_fingerprint,
                    "received": system_fingerprint
                }
            )
            return False, doc, "Invalid system fingerprint"

        return True, doc, None

    async def issue_key(
        self,
        raw_key: str,
        user_id: str,
        organization_id: str | None = None,
        operator_id: str | None = None,
        client_name: str = "operator",
        permissions: list[str] | None = None,
        status: ApiKeyStatus = ApiKeyStatus.ACTIVE,
    ) -> bool:
        """Issue (create and store) a new API key."""
        try:
            doc_id = self.make_doc_id(raw_key)
            doc_data = {
                "user_id": user_id,
                "organization_id": organization_id,
                "operator_id": operator_id,
                "client_name": client_name,
                "permissions": permissions or [],
                "status": status,
                "created_at": now(),
                "last_used_at": None,
                "expires_at": None,
            }

            result = await self.cache.create_document(
                collection=self.collection,
                document_id=doc_id,
                data=doc_data,
            )

            if result.success:
                logger.info(
                    "[API-KEY-SERVICE] API key issued",
                    extra={"doc_id": doc_id[:8] + "...", "user_id": user_id}
                )
                return True
            logger.error("[API-KEY-SERVICE] Failed to issue API key: %s", result.error)
            return False

        except Exception as e:
            logger.error("[API-KEY-SERVICE] Failed to issue API key: %s", e)
            return False

    async def revoke_key(self, raw_key: str) -> bool:
        """Mark an API key as REVOKED in the api_keys collection.

        Returns True on success or if the key did not exist (idempotent).
        Returns False only on storage failure.
        """
        if not raw_key:
            return True
        doc_id = self.make_doc_id(raw_key)
        try:
            existing = await self.cache.get_document_with_cache(self.collection, doc_id)
            if not existing:
                return True
            await self.cache.update_document(
                collection=self.collection,
                document_id=doc_id,
                data={"status": ApiKeyStatus.REVOKED, "revoked_at": now()},
                merge=True,
            )
            logger.info(
                "[API-KEY-SERVICE] API key revoked",
                extra={"doc_id": doc_id[:8] + "..."},
            )
            return True
        except Exception as e:
            logger.error("[API-KEY-SERVICE] Failed to revoke API key: %s", e)
            return False

    async def issue_operator_key(
        self,
        raw_key: str,
        user_id: str,
        organization_id: str | None,
        operator_id: str,
        settings_service: SettingsServiceProtocol,
        client_name: str = "operator",
        permissions: list[str] | None = None,
    ) -> bool:
        """Issue an operator API key."""
        return await self.issue_key(
            raw_key=raw_key,
            user_id=user_id,
            organization_id=organization_id,
            operator_id=operator_id,
            client_name=client_name,
            permissions=permissions,
            status=ApiKeyStatus.ACTIVE,
        )

    async def rotate_operator_key(
        self,
        old_raw_key: str | None,
        new_raw_key: str,
        user_id: str,
        organization_id: str | None,
        operator_id: str,
        settings_service: SettingsServiceProtocol,
        client_name: str = "operator",
        permissions: list[str] | None = None,
    ) -> bool:
        """Rotate an operator API key.

        Issues the new key first so a failure leaves the OLD key authoritative.
        Only after the new key is fully in place is the old key revoked.
        """
        ok = await self.issue_operator_key(
            raw_key=new_raw_key,
            user_id=user_id,
            organization_id=organization_id,
            operator_id=operator_id,
            settings_service=settings_service,
            client_name=client_name,
            permissions=permissions,
        )
        if not ok:
            return False

        if old_raw_key and old_raw_key != new_raw_key:
            revoked = await self.revoke_key(old_raw_key)
            if not revoked:
                logger.warning(
                    "[API-KEY-SERVICE] Old operator API key revocation failed "
                    "after successful rotation; reconciliation will catch up",
                    extra={"operator_id": operator_id},
                )

        return True

    async def revoke_operator_key(
        self,
        raw_key: str,
        settings_service: SettingsServiceProtocol,
    ) -> bool:
        """Revoke an operator API key in the canonical store."""
        return await self.revoke_key(raw_key)

    async def record_usage(self, raw_key: str, system_fingerprint: str | None = None) -> None:
        """Update the last used timestamp of a key and establish fingerprint if missing."""
        try:
            doc_id = self.make_doc_id(raw_key)
            data = await self.cache.get_document_with_cache(self.collection, doc_id)
            if not data:
                return

            doc = ApiKeyDocument.model_validate(data)
            updates: dict[str, Any] = {"last_used_at": now()}

            # Establish fingerprint if not already set (immutable thereafter)
            if not doc.system_fingerprint and system_fingerprint:
                updates["system_fingerprint"] = system_fingerprint
                logger.info(
                    "[API-KEY-SERVICE] Established system fingerprint for API key",
                    extra={
                        "doc_id": doc_id[:8] + "...",
                        "system_fingerprint": system_fingerprint
                    }
                )

            await self.cache.update_document(
                collection=self.collection,
                document_id=doc_id,
                data=updates,
                merge=True
            )
        except Exception as e:
            logger.warning("[API-KEY-SERVICE] Failed to record usage: %s", e)
